File: 437. Path Sum III.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def toleaf(self, node, path, sum):
        if not node:
            return 0
        
        ans = 0
        path.append(0)
        
        for i in range(len(path)):
            path[i] += node.val
            
            if path[i]==sum:
                ans += 1
                
        return ans+self.toleaf(node.left, path[:], sum)+self.toleaf(node.right, path[:], sum)
        
    def pathSum(self, root: TreeNode, sum: int) -> int:
        return self.toleaf(root, [], sum)

File: 437. Path Sum III.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} sum
 * @return {number}
 */
function toleaf(node, path, sum){
    if(!node){
        return 0
    }
    let ans = 0;
    path.push(0);

    for(let i=0;i<path.length;i++){
        path[i] += node.val;

        if(path[i]==sum){
            ans += 1;
        }
    }
    return ans+toleaf(node.left, Array.from(path), sum)+toleaf(node.right, Array.from(path), sum);
}
var pathSum = function(root, sum) {
    return toleaf(root, [], sum);
};